homeassistant/components/nmap_tracker/__init__.py

```diff
@@ -1 +1,372 @@
+"""The Nmap Tracker integration."""
+from __future__ import annotations
+
+import asyncio
+import contextlib
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+import logging
+
+import aiohttp
+from getmac import get_mac_address
+from mac_vendor_lookup import AsyncMacLookup
+from nmap import PortScanner, PortScannerError
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_EXCLUDE, CONF_HOSTS, EVENT_HOMEASSISTANT_STARTED
+from homeassistant.core import CoreState, HomeAssistant, callback
+from homeassistant.helpers import entity_registry as er
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.device_registry import format_mac
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.event import async_track_time_interval
+import homeassistant.util.dt as dt_util
+
+from .const import (
+    CONF_HOME_INTERVAL,
+    CONF_OPTIONS,
+    DOMAIN,
+    NMAP_TRACKED_DEVICES,
+    PLATFORMS,
+    TRACKER_SCAN_INTERVAL,
+)
+
+# Some version of nmap will fail with 'Assertion failed: htn.toclock_running == true (Target.cc: stopTimeOutClock: 503)\n'
+NMAP_TRANSIENT_FAILURE = "Assertion failed: htn.toclock_running == true"
+MAX_SCAN_ATTEMPTS = 16
+OFFLINE_SCANS_TO_MARK_UNAVAILABLE = 3
+
+
+def short_hostname(hostname):
+    """Return the first part of the hostname."""
+    if hostname is None:
+        return None
+    return hostname.split(".")[0]
+
+
+def human_readable_name(hostname, vendor, mac_address):
+    """Generate a human readable name."""
+    if hostname:
+        return short_hostname(hostname)
+    if vendor:
+        return f"{vendor} {mac_address[-8:]}"
+    return f"Nmap Tracker {mac_address}"
+
+
+@dataclass
+class NmapDevice:
+    """Class for keeping track of an nmap tracked device."""
+
+    mac_address: str
+    hostname: str
+    name: str
+    ipv4: str
+    manufacturer: str
+    reason: str
+    last_update: datetime.datetime
+    offline_scans: int
+
+
+class NmapTrackedDevices:
+    """Storage class for all nmap trackers."""
+
+    def __init__(self) -> None:
+        """Initialize the data."""
+        self.tracked: dict = {}
+        self.ipv4_last_mac: dict = {}
+        self.config_entry_owner: dict = {}
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Nmap Tracker from a config entry."""
+    domain_data = hass.data.setdefault(DOMAIN, {})
+    devices = domain_data.setdefault(NMAP_TRACKED_DEVICES, NmapTrackedDevices())
+    scanner = domain_data[entry.entry_id] = NmapDeviceScanner(hass, entry, devices)
+    await scanner.async_setup()
+    entry.async_on_unload(entry.add_update_listener(_async_update_listener))
+    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
+    return True
+
+
+async def _async_update_listener(hass: HomeAssistant, entry: ConfigEntry):
```

Comment:
Missing return value typing.

Reply:
This feature/PR has been fully reverted during the beta, as it locked up Home Assistant as a whole.

Reply:
Right, forgot about that. 👍

Reply:
fixed in https://github.com/home-assistant/core/pull/54715